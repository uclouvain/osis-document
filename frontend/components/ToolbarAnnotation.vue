<!--
  -
  - OSIS stands for Open Student Information System. It's an application
  - designed to manage the core business of higher education institutions,
  - such as universities, faculties, institutes and professional schools.
  - The core business involves the administration of students, teachers,
  - courses, programs and so on.
  -
  - Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  - This program is free software: you can redistribute it and/or modify
  - it under the terms of the GNU General Public License as published by
  - the Free Software Foundation, either version 3 of the License, or
  - (at your option) any later version.
  -
  - This program is distributed in the hope that it will be useful,
  - but WITHOUT ANY WARRANTY; without even the implied warranty of
  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  - GNU General Public License for more details.
  -
  - A copy of this license - GNU General Public License - is available
  - at the root of the source code of this program.  If not,
  - see http://www.gnu.org/licenses/.
  -
  -->

<template>
  <div class="btn-group dropup annotation-dropdown">
    <button
        class="btn"
        type="button"
        :class="isActivated ? 'btn-primary': 'btn-default'"
        @click="$emit( 'onChange', isActivated ? '': color || Object.values(colors)[0] )"
    >
      <span
          class="fas"
          :class="icon"
      />
    </button>
    <button
        type="button"
        class="btn btn-default dropdown-toggle"
        data-toggle="dropdown"
    >
      <span class="caret" />
    </button>
    <ul class="dropdown-menu">
      <li
          v-for="(rgb, color_name) in colors"
          :key="color_name"
          :class="`${classPrefix}-${color_name}`"
      >
        <a
            href="#"
            :class="`${classPrefix}-${color_name}`"
            @click.prevent="$emit('onChange', rgb)"
        >
          {{ $t(`editor.colors.${color_name}`) }}
        </a>
      </li>
    </ul>
  </div>
</template>

<script setup lang="ts">
defineProps<{
  isActivated: boolean,
  color: string,
  classPrefix: string,
  icon: string,
  colors: Record<string, string>,
}>();
defineEmits<{
  (e: 'onChange', color: string): void,
}>();
</script>

<style lang="scss">
.annotation-dropdown .dropdown-menu > li > a {
  color: inherit;
}
</style>
