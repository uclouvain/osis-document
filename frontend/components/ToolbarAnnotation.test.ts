/*
 *
 * OSIS stands for Open Student Information System. It's an application
 * designed to manage the core business of higher education institutions,
 * such as universities, faculties, institutes and professional schools.
 * The core business involves the administration of students, teachers,
 * courses, programs and so on.
 *
 * Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of this license - GNU General Public License - is available
 * at the root of the source code of this program.  If not,
 * see http://www.gnu.org/licenses/.
 *
 */

import {it, expect} from 'vitest';
import {mount} from "@vue/test-utils";
import ToolbarAnnotation from "./ToolbarAnnotation.vue";

const props = {
  isActivated: false,
  color: '',
  classPrefix: 'bg',
  icon: 'hightlighy',
  colors: {'danger': '#ff0000', 'warning': '#ffff00'},
};

it('should mount', () => {
  const wrapper = mount(ToolbarAnnotation, {props});
  expect(wrapper.html()).toMatchSnapshot();
});

it('should emit default color when not activated', async () => {
  const wrapper = mount(ToolbarAnnotation, {props});
  await wrapper.find('.btn-default:not(.dropdown-toggle)').trigger('click');
  expect(wrapper.emitted()).toHaveProperty('onChange.0.0', '#ff0000');
});

it('should emit no color when activated', async () => {
  const wrapper = mount(ToolbarAnnotation, {
    props: {
      ...props,
      isActivated: true,
    },
  });
  await wrapper.find('.btn-primary').trigger('click');
  expect(wrapper.emitted()).toHaveProperty('onChange.0.0', '');
});

it('should emit chosen color', async () => {
  const wrapper = mount(ToolbarAnnotation, {props});
  await wrapper.find('a.bg-warning').trigger('click');
  expect(wrapper.emitted()).toHaveProperty('onChange.0.0', '#ffff00');
});
