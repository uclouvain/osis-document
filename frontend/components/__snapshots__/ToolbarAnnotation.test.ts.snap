// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should mount 1`] = `
"<div class=\\"btn-group dropup annotation-dropdown\\"><button class=\\"btn btn-default\\" type=\\"button\\"><span class=\\"fas hightlighy\\"></span></button><button type=\\"button\\" class=\\"btn btn-default dropdown-toggle\\" data-toggle=\\"dropdown\\"><span class=\\"caret\\"></span></button>
  <ul class=\\"dropdown-menu\\">
    <li class=\\"bg-danger\\"><a href=\\"#\\" class=\\"bg-danger\\">editor.colors.danger</a></li>
    <li class=\\"bg-warning\\"><a href=\\"#\\" class=\\"bg-warning\\">editor.colors.warning</a></li>
  </ul>
</div>"
`;
